src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -0,0 +1,319 @@
+			this.retryTemplate.registerListener(this.listener);
+		}
+		this.messageListenerContainer.getContainerProperties().setMessageListener(listener);
+	}
+
+	@Override
+	protected void doStart() {
+		this.messageListenerContainer.start();
+	}
+
+	@Override
+	protected void doStop() {
+		this.messageListenerContainer.stop();
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:inbound-gateway";
+	}
+
+	@Override
+	public int beforeShutdown() {
+		this.messageListenerContainer.stop();
+		return getPhase();
+	}
+
+	@Override
+	public int afterShutdown() {
+		return getPhase();
+	}
+
+	/**
+	 * If there's a retry template, it will set the attributes holder via the listener. If
+	 * there's no retry template, but there's an error channel, we create a new attributes
+	 * holder here. If an attributes holder exists (by either method), we set the
+	 * attributes for use by the {@link ErrorMessageStrategy}.
+	 * @param record the record.
+	 * @param message the message.
+	 */
+	private void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needHolder = getErrorChannel() != null && this.retryTemplate == null;
+		boolean needAttributes = needHolder | this.retryTemplate != null;
+		if (needHolder) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
+				attributes.setAttribute(KafkaHeaders.RAW_DATA, record);
+			}
+		}
+	}
+
+	@Override
+	protected AttributeAccessor getErrorMessageAttributes(Message<?> message) {
+		AttributeAccessor attributes = attributesHolder.get();
+		if (attributes == null) {
+			return super.getErrorMessageAttributes(message);
+		}
+		else {
+			return attributes;
+		}
+	}
+
+	private class IntegrationRecordMessageListener extends RecordMessagingMessageListenerAdapter<K, V>
+			implements RetryListener {
+
+		IntegrationRecordMessageListener() {
+			super(null, null);
+		}
+
+		@Override
+		public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment, Consumer<?, ?> consumer) {
+			Message<?> message = null;
+			try {
+				message = toMessagingMessage(record, acknowledgment, consumer);
+				if (KafkaInboundGateway.this.retryTemplate != null) {
+					message = addDeliveryAttemptHeader(message);
+				}
+				setAttributesIfNecessary(record, message);
+			}
+			catch (RuntimeException e) {
+				if (getErrorChannel() != null) {
+					KafkaInboundGateway.this.messagingTemplate.send(getErrorChannel(), buildErrorMessage(null,
+							new ConversionException("Failed to convert to message for: " + record, e)));
+				}
+			}
+			if (message != null) {
+				try {
+					Message<?> reply = sendAndReceiveMessage(message);
+					if (reply != null) {
+						reply = enhanceReply(message, reply);
+						KafkaInboundGateway.this.kafkaTemplate.send(reply);
```

Comment:
Gary, there is a failing test on Travis:
```
org.springframework.integration.kafka.inbound.InboundGatewayTests > testInboundErrorRecover FAILED
    org.junit.ComparisonFailure at InboundGatewayTests.java:216
```
And I somehow think that the problem that we don't call `flush()` here after `send()`.

WDYT?

Reply:
I doubt if it's flush - we wait for 60 seconds; I'll take a look.

Reply:
Yeah, might be not related.
Anyway how about to make all those `kafkaTemplate` as `autoFlush = true` to decrease the waiting for reply interval on the client side? Otherwise we send reply only by the flusher on background in the Kafka Client and that has some period as well.

Also  I think it might worth to mention in the docs that sending reply behavior depends directly of the `autoFlush` flag on the `KafkaTemplate`.

WDYT?

Reply:
ok; will push another

Reply:
Actually, no; yes we can change the test, but I don't think a general autoFlush recommendation is appropriate - since the gateway is async, there's no reason why they can't send multiple requests in the same ip packet. I will make a mention in the documentation, though.

Note that we need a separate PR against S-K for the docs.

Reply:
Are you still OK that we are talking here about an **Inbound** Gateway and replies, not requests? 

In my feeling that would be better to send reply to the waiting client as fast as possible. That's why I'd even prefer to have explicit `flush()` here in the `KafkaInboundGateway`.

Just IMHO...

Reply:
`linger.ms` is 0 by default so the flush is not needed: https://kafka.apache.org/documentation/#configuration

Reply:
>  This setting defaults to 0 (i.e. no delay).

We don't need that `autoFlush` at all - such a producer property does the trick 😄 

Thank you for your patience!
So, then yes: 10 secs to wait for reply on the slow Travis definitely might be as a bottleneck.

Reply:
`autoFlush` is still valid - they might have several templates using the same producer factory - I will clean up the javadocs though.